app/src/ingestion/pdf_postprocess.py

```diff
@@ -223,6 +223,31 @@ def _should_merge_text_split_across_pages(text: EnrichedText, next_text: Enriche
     return next_text.text.strip()[0].islower()
 
 
+NestedHeadingTitles = tuple[str, ...]
+
+
+def _group_headings_text(markdown_texts: list[EnrichedText]) -> list[EnrichedText]:
+    grouped_texts_by_headings: dict[NestedHeadingTitles, EnrichedText] = {}
+    for ind, markdown_text in enumerate(markdown_texts):
+        text_nested_headings: NestedHeadingTitles = tuple(
+            [h.title for h in markdown_texts[ind].headings]
+        )
+        if text_nested_headings in grouped_texts_by_headings:
+            grouped_texts_by_headings[text_nested_headings].text += f"\n{markdown_texts[ind].text}"
```

Comment:
Could we make this a double newline ("\n\n") so that the sections are separate, e.g.,:
```
This is some paragraph text.

This is another paragraph but preceding a list:
 - List item
 - Another list item

A new paragraph
```

rather than

```
This is some paragraph text.
This is another paragraph but preceding a list:
 - List item
 - Another list item
A new paragraph
```

Reply:
We might run into cases like this:
```
This is some paragraph text.

This is another paragraph but preceding a list:
 - List item
 
 - Another list item

A new paragraph```